### DISCUSSION TITLE
Enforcing Java Record Invariants With Bean Validation - Gunnar Morling

### DISCUSSION THREAD

**Discussion by blog-comment-importer** [gunnarmorling/discussions.morling.dev#271] [Announcements]
https://www.morling.dev/blog/enforcing-java-record-invariants-with-bean-validation/

**Comment by blog-comment-importer** [comment]
_**Ismail Marmoush** wrote at 2020-09-10T19:03:03Z (comment imported from Disqus):_

Actually you can just do the following with record, it's from <a href="https://openjdk.java.net/jeps/384" rel="nofollow noopener" title="https://openjdk.java.net/jeps/384">https://openjdk.java.net/je...</a>

```
public record Bob(String name) {
  public Bob {
    if (name.equals("blabla"))
      throw new IllegalArgumentException("bad name");
  }
}
```

**Reply by blog-comment-importer** [reply]
_**Gunnar Morling** wrote at 2020-09-10T19:47:09Z (comment imported from Disqus):_

<p>Of course you can do that, and there's nothing wrong with it. The suggested solution based on BV has some advantages over this, though, as described in the post; e.g. constraints become examinable metadata of the API, you have I18N support and more.</p>

**Comment by lostiniceland** [comment]
The benefit of BV is also that you get all validations in one result, which has the advantage that lets say you are replying to a user request the bespoke user gets all error he made at once and not one after the other.

Still, I am quite disappointed by many EE-Apis and their lack of proper record support (we just ditched JSONB in favor of Jackson because it was unbearable where Jackson just worked).

Using Records for DTOs for instance is great, because you can be sure nobody tampered with the data when it reaches your component. But you cannot do this when you need to put BV on them (without the ByteBuddy-Magic described here). Also you can forget to use a record as a value-object inside a JPA-entity when BV is needed.

From a user perspective, I think it would be great if BV would be able to validate a record after it was created (because if the validation is triggered within the constructor BV would only return errors for this records but not the ones which might be placed on the enclosing object).
Consider the example mentioned earlier where a records is used as a value-object within an entity. The record should be able to be created from invalid (user-)data but when BV is called from JPA the invalid record-data must be reported.

**Comment by tstavinoha** [comment]
OK so this is really cool, but the workaround for type-level constraints is a real bummer.
I've fiddled around with the code a bit and I think I got it to work with type-level annotations by doing the following changes:

ValiodationInterceptor:
```
 public static <T> void validate(@This T object) {
        Set<ConstraintViolation<T>> violations = validator.validate(object);
```

clean + compile afterwards.

I'm wondering if you have already tried this and/or see any possible issues with this approach?

**Reply by tstavinoha** [reply]
record example:

```
@ValidDurationTimes
public record Duration(@NotNull Instant startTime,
                       @NotNull Instant endTime) {

}
```

**Reply by joseph-pachod** [reply]
as far as i've seen: in this case the `@ParameterScriptAssert` isn't triggered. 

but i prefer bean level validation than this, so fine by me personally

**Comment by ybermejo** [comment]
Hi,

I am new with this Java Records, can someone help me with my issue?

I have 2 Records (PointsRedemptionV2Body & RedemptionItemBody) and I declared RedemptionItemBody as an array in the first record but the problem is the validation constraints from RedemptionItemBody are not working.

public record PointsRedemptionV2Body(

        String orderReferenceNumber,

        @NotNull
        List<RedemptionItemBody> redemptionItemList) {

}


public record RedemptionItemBody(

        @Min(value = 1L)
        Integer redemptionItemNumber,

        @Min(value = 1L)
        Integer quantity,

        @NotBlank
        String awardId,

        String catalogId,

        @NotBlank
        String shortDescription,

        @NotBlank
        String longDescription,

        @NotBlank
        String awardType,

        @Min(value = 0L)
        @MaxPointQuantity
        Integer awardPointCost,

        @NotNull
        @Positive
        Float companyCost
        ) {

}

**Reply by ybermejo** [reply]
Hi @Gunnar Morling, can you help me?
